crates/shim/src/lib.rs

```diff
@@ -291,21 +294,36 @@ fn handle_signals(mut signals: Signals) {
                     return;
                 }
                 SIGCHLD => loop {
-                    unsafe {
-                        let pid: pid_t = -1;
-                        let mut status: c_int = 0;
-                        let options: c_int = libc::WNOHANG;
-                        let res_pid = libc::waitpid(pid, &mut status, options);
-                        let status = libc::WEXITSTATUS(status);
-                        if res_pid <= 0 {
+                    // If WUNTRACED is not set, both suspended and running process will
+                    // bring WaitStatus::StillAlive and we cannot distinguish them.
+                    // Note that this thread sticks to child even it is suspended.
+                    // (it will repeatedly bring WaitStatus::StillAlive after once WaitStatus::Stopped is reported)
+                    match wait::waitpid(
+                        Some(Pid::from_raw(-1)),
+                        Some(WaitPidFlag::WNOHANG | WaitPidFlag::WUNTRACED),
+                    ) {
+                        Ok(WaitStatus::Exited(pid, status)) => {
+                            monitor_notify_by_pid(pid.as_raw(), status)
+                                .unwrap_or_else(|e| error!("failed to send exit event {}", e))
+                        }
+                        Ok(WaitStatus::Signaled(pid, sig, _)) => {
+                            // TODO: add signal notification for monitor
+                            warn!("child {} terminated({})", pid, sig);
+                        }
+                        Ok(WaitStatus::Stopped(pid, sig)) => {
+                            warn!("child {} stopped({})", pid, sig);
```

Comment:
Should we just ignore these since they may be too many and thay are not representing "warning" messages.
( You can test on a real container to see how many they are )

Reply:
Fine.
I removed, thank you !